### PR TITLE
Add move detection and support to ASLayoutTransition

### PR DESCRIPTION
**Pull request by wiseoldduck**
...and NSArray+Diffing.
Add some tests.

See https://github.com/TextureGroup/Texture/pull/816 for more details on the existing shortcomings. This PR should allow z-order to be correct after nodes move during a transition.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASLayout+IGListKit.h:</p>
<pre><code>//
//  ASLayout+IGListKit.h
//  Texture
//
//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASLayout+IGListKit.mm:</p>
<pre><code>//
//  ASLayout+IGListKit.mm
//  Texture
//
//  Copyright (c) 2018-present, Pinterest, Inc.  All rights reserved.
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Details/NSArray+Diffing.h [hunk 1]
```diff
@@ -35,4 +35,12 @@
  */
 - (void)asdk_diffWithArray:(NSArray *)array insertions:(NSIndexSet **)insertions deletions:(NSIndexSet **)deletions compareBlock:(BOOL (^)(id lhs, id rhs))comparison;
 
+/**
+ * @abstract Compares two arrays, providing the insertion, deletion, and move indexes needed to transform into the target array.
+ * @discussion This compares the equality of each object with `isEqual:`.
```
Comment (appleguy) [inline Source/Details/NSArray+Diffing.h]: I wonder if we should add to the method above that it will not transform into the same order, only to the same contents?

File: Source/Details/NSArray+Diffing.h [hunk 2]
```diff
@@ -35,4 +35,12 @@
...
+ * This diffing algorithm uses a bottom-up memoized longest common subsequence solution to identify differences.
+ * It runs in O(mn) complexity.
+ * The moves are returned in ascending order of their destination index.
```
Comment (appleguy) [inline Source/Details/NSArray+Diffing.h]: Usually, the moves are returned in a form that applies either before any changes, after insertions, or after deletions - in other words, the order of insert / delete being applied (if at all) does matter for the move indices being valid.

Could you expand a bit on the comment to describe the exact order of application for insertions, deletions, and moves that results in the correct transformation?

File: Source/Details/NSArray+Diffing.mm [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                     moves:(NSArray<NSIndexPath *> **)moves compareBlock:(compareBlock)comparison
+{
+  typedef std::unordered_map<NSUInteger, std::queue<NSUInteger>> move_map;
+  NSAssert(comparison != nil, @"Comparison block is required");
+    NSAssert(moves == nil || comparison == [NSArray defaultCompareBlock], @"move detection requires isEqual: and hash (no custom compare)");
+  std::unique_ptr<move_map> potentialMoves(nullptr);
+  NSMutableArray<NSIndexPath *> *moveIndexPaths = nil;
+  NSMutableIndexSet *insertionIndexes = nil, *deletionIndexes = nil;
+  if (moves) {
+    potentialMoves = std::unique_ptr<move_map>(new move_map());
```
Comment (appleguy) [inline Source/Details/NSArray+Diffing.mm]: Can you change this to stack allocation, matching the style of other unordered_map's in Texture? I think we can skip the typedef in that case too, as it is fairly clear in that style.

This will allocate it in some cases where it is not needed, but as a stack allocation, it is quite efficient.

File: Source/Details/NSArray+Diffing.mm [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        (*potentialMoves)[hash].pop();
+        if ((*potentialMoves)[hash].empty()) {
+          potentialMoves->erase(hash);
+        }
+        if (movedFrom != j) {
+          NSUInteger indexes[] = {movedFrom, j};
+          [moveIndexPaths addObject:[NSIndexPath indexPathWithIndexes:indexes length:2]];
+        }
+      }
+        if (i < commonObjects.count && j < array.count && comparison(commonObjects[i], array[j])) {
```
Comment (Adlai-Holler) [inline Source/Details/NSArray+Diffing.mm]: Indentation here

File: Source/Private/ASLayoutTransition.mm [hunk 5]
```diff
@@ -254,7 +300,7 @@ - (ASSizeRange)transitionContext:(_ASTransitionContext *)context constrainedSize
   for (ASLayout *sublayout in layout.sublayouts) {
     if (idx > lastIndex) { break; }
     if (idx >= firstIndex && [indexes containsIndex:idx]) {
-      ASDisplayNode *node = (ASDisplayNode *)sublayout.layoutElement;
+      ASDisplayNode *node = ASDynamicCast(sublayout.layoutElement, ASDisplayNode);
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: Is this change necessary? Not that this is a particularly hot path but `ASDynamicCast` incurs a runtime cost whereas a static cast is free (if you know it's safe.)

File: Source/Private/ASLayoutTransition.mm [hunk 6]
```diff
@@ -156,18 +180,40 @@ - (void)calculateSubnodeOperationsIfNeeded
...
 
   if (previousLayout) {
+#if AS_IG_LIST_KIT
+    // IGListDiff completes in linear time O(m+n), so use it if we have it:
+    IGListIndexSetResult *result = IGListDiff(previousLayout.sublayouts, pendingLayout.sublayouts, IGListDiffEquality);
+    _insertedSubnodePositions = findNodesInLayoutAtIndexes(pendingLayout, result.inserts, &_insertedSubnodes);
+    _removedSubnodePositions = findNodesInLayoutAtIndexes(previousLayout, result.deletes, &_removedSubnodes);
+    for (IGListMoveIndex *move in result.moves) {
+      id subnode = previousLayout.sublayouts[static_cast<NSUInteger>(move.from)].layoutElement;
+      _subnodeMoves.push_back(std::pair<id, NSUInteger>(subnode, move.to));
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: - A C-cast `((NSUInteger)move.from)` is probably sufficient instead of a `static_cast` (which is usually used for C++ types). Is any kind of cast actually necessary though – the implicit conversion shouldn't be dangerous.
- This is a good place to use `unowned id subnode` to avoid the extra retain/release pair that ARC will insert otherwise – we will retain the object into the `std::pair` on the next line so it's safe.

File: Source/Private/ASLayoutTransition.mm [hunk 7]
```diff
@@ -67,6 +73,7 @@ @implementation ASLayoutTransition {
   NSArray<ASDisplayNode *> *_removedSubnodes;
   std::vector<NSUInteger> _insertedSubnodePositions;
   std::vector<NSUInteger> _removedSubnodePositions;
+  std::vector<std::pair<id, NSUInteger>> _subnodeMoves;
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: Let's use `ASDisplayNode *` here for some type-safety.

File: Source/Private/ASLayoutTransition.mm [hunk 8]
```diff
@@ -98,27 +105,44 @@ - (BOOL)isSynchronous
...
+  if (_insertedSubnodePositions.size() == 0 && _subnodeMoves.size() == 0) {
     return;
   }
 
   ASDisplayNodeLogEvent(_node, @"insertSubnodes: %@", _insertedSubnodes);
   NSUInteger i = 0;
-  for (ASDisplayNode *node in _insertedSubnodes) {
+  NSUInteger j = 0;
+  for (j = 0; j < _subnodeMoves.size(); ++j) {
+    [_subnodeMoves[j].first _removeFromSupernodeIfEqualTo:_node];
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: I think we can do `for (auto const &move : _subnodeMoves) { [move.first _remove…] }` and use C++ enumeration below (during the insertions) as well.

File: Source/Private/ASLayoutTransition.mm [hunk 9]
```diff
@@ -156,18 +180,40 @@ - (void)calculateSubnodeOperationsIfNeeded
...
+
     _insertedSubnodePositions = findNodesInLayoutAtIndexes(pendingLayout, insertions, &_insertedSubnodes);
     _removedSubnodePositions = findNodesInLayoutAtIndexesWithFilteredNodes(previousLayout,
                                                                            deletions,
                                                                            _insertedSubnodes,
                                                                            &_removedSubnodes);
+    // These should arrive sorted in ascending order of move destinations.
+    for (NSIndexPath *move in moves) {
+      id subnode = previousLayout.sublayouts[static_cast<NSUInteger>([move indexAtPosition:0])].layoutElement;
+      _subnodeMoves.push_back(std::pair<id, NSUInteger>(subnode, [move indexAtPosition:1]));
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: Is a cast here necessary? NSArray indexing should do the conversion implicitly.

File: Source/Details/NSArray+Diffing.h (hunk 1 above)
Comment (wiseoldduck) [reply Source/Details/NSArray+Diffing.h]: ahh but it will, you only need the moves if you want to animate them

File: Source/Private/ASLayoutTransition.mm (hunk 5 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: Haha this has nothing to do with the rest of this does it. But I noticed the comment below which was was claiming a behavior that simple type casting doesn't actually have!

Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: // Ignore the odd case in which a non-node sublayout is accessed and the type cast fails
      if (node != nil) {

File: Source/Private/ASLayoutTransition.mm (hunk 9 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: Same as above, it was for clarity, but I don't even agree with past-me about this one - it already is a NSUInteger!

File: Source/Details/NSArray+Diffing.h (hunk 2 above)
Comment (wiseoldduck) [reply Source/Details/NSArray+Diffing.h]: Will do - these are returned as simple (previousIndex -> newIndex) moves, which is easy to work with and consistent with what vanilla `[UICollectionView performBatchUpdates:completion:]` wants

File: Source/Details/NSArray+Diffing.mm [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      }
+
+    }
+  }
+
+  if (insertions || moves) {
+    insertionIndexes = [NSMutableIndexSet indexSet];
+    NSArray *commonObjects = [self objectsAtIndexes:commonIndexes];
+    BOOL moveFound;
+    NSUInteger movedFrom = NSNotFound;
```
[the remark this answers is not in the document]
Comment (wiseoldduck) [reply Source/Details/NSArray+Diffing.mm]: Definitely can go in the loop. The BOOL was added in a vain attempt to make this more readable lol

**Comment by wiseoldduck** [comment]
Thanks guys, the comments I didn't reply to are all "Will do"s!

File: Source/Private/ASLayoutTransition.mm (hunk 6 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: The cast is entirely for "clarity", but I don't really know why I thought it helped :-| :D

File: Source/ASDisplayNode+Layout.mm [hunk 11]
```diff
@@ -680,7 +680,7 @@ - (void)transitionLayoutWithSizeRange:(ASSizeRange)constrainedSize
       }
       
       // Apply the subnode insertion immediately to be able to animate the nodes
-      [pendingLayoutTransition applySubnodeInsertions];
+      [pendingLayoutTransition applySubnodeInsertionsAndMoves];
```
[the remark this answers is not in the document]
Comment (wiseoldduck) [reply Source/ASDisplayNode+Layout.mm]: That's great, maybe it is a cross-fade animation or something. But yes being able to animate that is a good win from this.
The second point is very related, as setting up animations that will run simultaneously also allows us to do the setup in whatever order makes life easiest for ourselves,  not necessarily tied to the order in which anything plays out on the screen. But at present (where we are not animating z-order) I agree it looks nicer to have the z-order swap happen before the animations rather than at the end.

File: Source/Details/NSArray+Diffing.h (hunk 1 above)
Comment (wiseoldduck) [reply Source/Details/NSArray+Diffing.h]: I will add some comments on this, as it was not clear to me until close to the end that this was actually true (that the -/+ information is sufficient and accurate to do the reordering without moves). It requires applying the diff in a certain way (basically in order into a sparse array) that was neither documented nor being done.

File: Source/Private/ASLayoutTransition.mm (hunk 5 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: I will move this into a separate #trivial PR as there are already plenty of moving parts here to keep straight ;)

File: Source/Details/NSArray+Diffing.mm [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+      moveFound = (potentialMoves && potentialMoves->count(hash));
+      if (moveFound) {
+        movedFrom = (*potentialMoves)[hash].front();
+        (*potentialMoves)[hash].pop();
+        if ((*potentialMoves)[hash].empty()) {
+          potentialMoves->erase(hash);
+        }
+        if (movedFrom != j) {
+          NSUInteger indexes[] = {movedFrom, j};
+          [moveIndexPaths addObject:[NSIndexPath indexPathWithIndexes:indexes length:2]];
```
[the remark this answers is not in the document]
Comment (wiseoldduck) [reply Source/Details/NSArray+Diffing.mm]: Ummmmm yeah that's exactly what I was thinking...:p

Haha no I had not `.mm`-ed the file yet and was in need of something like a `std::pair` of Indexes. There were all these `NSIndexSet`s floating around, and `NSIndexPath` jumped right to mind. It seems more natural to me, now, to go ahead and switch to a `pair`, but this tagged pointer detail is very interesting! I will leave it as is for now and educate myself on tagged pointers ;)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASLayoutTransition.mm (hunk 8 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: I don't think it is so straightforward below (insertions/moves), as we are interleaving the two vectors and consuming from only one or the other each time.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASLayoutTransition.mm [hunk 13]
```diff
@@ -156,18 +180,40 @@ - (void)calculateSubnodeOperationsIfNeeded
...
 
   if (previousLayout) {
+#if AS_IG_LIST_KIT
+    // IGListDiff completes in linear time O(m+n), so use it if we have it:
+    IGListIndexSetResult *result = IGListDiff(previousLayout.sublayouts, pendingLayout.sublayouts, IGListDiffEquality);
+    _insertedSubnodePositions = findNodesInLayoutAtIndexes(pendingLayout, result.inserts, &_insertedSubnodes);
+    _removedSubnodePositions = findNodesInLayoutAtIndexes(previousLayout, result.deletes, &_removedSubnodes);
+    for (IGListMoveIndex *move in result.moves) {
+      unowned ASDisplayNode *subnode = previousLayout.sublayouts[move.from].layoutElement;
+      _subnodeMoves.push_back(std::pair<ASDisplayNode *, NSUInteger>(subnode, move.to));
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: Import `<utility>` and use `std::make_pair(subnode, move.to)`. This:

- Uses type inference to remove code.
- Makes the pair `<unowned ASDisplayNode *, NSUInteger>`, skipping the extra retain/release.

File: Source/Private/ASLayoutTransition.mm [hunk 14]
```diff
@@ -156,18 +180,40 @@ - (void)calculateSubnodeOperationsIfNeeded
...
+    }
+
+    // Sort by ascending order of move destinations, this will allow easy loop of `insertSubnode:AtIndex` later.
+    std::sort(_subnodeMoves.begin(), _subnodeMoves.end(), [](std::pair<id<ASLayoutElement>, NSUInteger> a,
+            std::pair<ASDisplayNode *, NSUInteger> b) {
+      return a.second < b.second;
+    });
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: Use `auto a, auto b` to save space

File: Source/Private/ASLayoutTransition.mm [hunk 15]
```diff
@@ -156,18 +180,40 @@ - (void)calculateSubnodeOperationsIfNeeded
...
+
     _insertedSubnodePositions = findNodesInLayoutAtIndexes(pendingLayout, insertions, &_insertedSubnodes);
     _removedSubnodePositions = findNodesInLayoutAtIndexesWithFilteredNodes(previousLayout,
                                                                            deletions,
                                                                            _insertedSubnodes,
                                                                            &_removedSubnodes);
+    // These should arrive sorted in ascending order of move destinations.
+    for (NSIndexPath *move in moves) {
+      unowned ASDisplayNode *subnode = previousLayout.sublayouts[([move indexAtPosition:0])].layoutElement;
+      _subnodeMoves.push_back(std::pair<id, NSUInteger>(subnode, [move indexAtPosition:1]));
```
Comment (Adlai-Holler) [inline Source/Private/ASLayoutTransition.mm]: - Is it valid to remove the parens on 213? `previousLayout.sublayouts[[move indexAtPosition:0]].layoutElement`
- Same `make_pair` game to safe space and retain/release traffic

File: Source/Layout/ASLayout+IGListKit.mm [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+- (BOOL)isEqualToDiffableObject:(nullable id <IGListDiffable>)other
+{
+  if (other == self) return YES;
+
+  ASLayout *otherLayout = ASDynamicCast(other, ASLayout);
+  if (!otherLayout) return NO;
+
+  return [otherLayout.layoutElement isEqual:self.layoutElement];
```
Comment (Adlai-Holler) [inline Source/Layout/ASLayout+IGListKit.mm]: Nit: remove nullability annotations from implementation file to be consistent with the rest of the framework style.

If it's easy to do, import the header that exposes the `_layoutElement` instance variable and use that directly to avoid extra retain/release on these elements. If not, it's not a blocker for this diff.

File: Source/Private/ASLayoutTransition.mm (hunk 13 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: Fascinating! So it will infer `unowned`? Is that a general rule for C++ type inference of ObjC types?

Comment (Adlai-Holler) [reply Source/Private/ASLayoutTransition.mm]: Yep!

File: Source/Private/ASLayoutTransition.mm (hunk 14 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: "Generic lambdas are not supported by the compiler"

Comment (Adlai-Holler) [reply Source/Private/ASLayoutTransition.mm]: What? That's crazy, it works in a sample project! OK must be something contextual, no worries

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Private/ASLayoutTransition.mm (hunk 14 above)
Comment (wiseoldduck) [reply Source/Private/ASLayoutTransition.mm]: that's interesting, do you guys ever build from the .xcworkspace/xcodeproj? I don't know much about BUCK but it seems like I have seen some other discrepancies

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/82d096d04781fa2b8bc2a031c9f50cec/raw/fda7b2571b79b635fc9acd334c0206a1a8297ce4/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/b22fd65cd9bc7d75ebdb221342942381/raw/ba5e3a3fbd4a476a44cfcb4905602f0d4fd53b5e/log.txt)

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Layout/ASLayout+IGListKit.h [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#if AS_IG_LIST_KIT
+#import <AsyncDisplayKit/ASLayout.h>
+#import <IGListKit/IGListKit.h>
+@interface ASLayout(IGListKit) <IGListDiffable>
+@end
+
+#endif // AS_IG_LIST_KIT
```
Comment (Adlai-Holler) [inline Source/Layout/ASLayout+IGListKit.h]: Now that we're diffing nodes, not layouts, can we remove this?

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Layout/ASLayout+IGListKit.h (hunk 17 above)
Comment (Adlai-Holler) [reply Source/Layout/ASLayout+IGListKit.h]: We talked on Slack, moving the IGListKit approach to diff node-arrays rather than layouts could come in a separate diff.
